README.md
Cabinet
=======

[![Build Status](https://travis-ci.org/dotsunited/Cabinet.svg?branch=master)](http://travis-ci.org/dotsunited/Cabinet)

Cabinet is a PHP 5.3+ library providing a simple file storage layer.

It provides a unified API for storing and retrieving files as well as getting basic informations like size and content type of a file.

This is useful if

  * you want to write reusable components with a configurable file storage backend
  * you want to ensure scalability for your file storage (You can for example start storing your files on your servers filesystem and switch later to Amazon S3)

Cabinet offers adapters for [PHP streams](http://php.net/stream) and [Amazon S3](https://s3.amazonaws.com) out of the box. But you can easily write your own adapters by implementing `DotsUnited\Cabinet\Adapter\AdapterInterface`.

Installation
------------

Cabinet can be installed using the [Composer](http://packagist.org) tool. You can either add `dotsunited/cabinet` to the dependencies in your composer.json, or if you want to install Cabinet as standalone, go to the main directory and run:

```bash
$ wget http://getcomposer.org/composer.phar
$ php composer.phar install
```

You can then use the composer-generated autoloader to access the Cabinet classes:

```php
<?php
require 'vendor/autoload.php';
?>
```

Usage
-----

Instances of Cabinet adapters can be either created directly or using the static `DotsUnited\Cabinet\Cabinet::factory()` method.

### Using a Cabinet Adapter constructor ###

You can create an instance of an adapter using its constructor. An adapter constructor takes one argument, which is an array of configuration parameters.

```php
<?php
$adapter = new \DotsUnited\Cabinet\Adapter\StreamAdapter(array(
    'base_path' => '/my/base/path',
    'base_uri'  => 'http://static.example.com'
));
?>
```

### Using the Cabinet factory ###

As an alternative to using an adapter constructor directly, you can create an instance of an adapter using the static method `DotsUnited\Cabinet\Cabinet::factory()`.

The first argument is a string that names the adapter class (for example '\DotsUnited\Cabinet\Adapter\StreamAdapter'). The second argument is the same array of parameters you would have given to the adapter constructor.

```php
<?php
$adapter = \DotsUnited\Cabinet\Cabinet::factory('\DotsUnited\Cabinet\Adapter\StreamAdapter', array(
    'base_path' => '/my/base/path',
    'base_uri'  => 'http://static.example.com'
));
?>
```

Alternatively, the first argument can be an associative array. The adapter class is then read from the 'adapter' key. Optionally, it can contain a 'config' key holding the configuration parameters. In this case, the second argument will be ignored.

```php
<?php
$adapter = \DotsUnited\Cabinet\Cabinet::factory(array(
    'adapter' => '\DotsUnited\Cabinet\Adapter\StreamAdapter',
    'config' => array(
        'base_path' => '/my/base/path',
        'base_uri'  => 'http://static.example.com'
    )
));
?>
```

### Managing files ###

Once you created the adapter, you can store, retrieve and get informations about files with the following methods:

#### Import a external local file: ####
```php
<?php
$adapter->import($external, $file);
?>
```

#### Write data to a file: ####
```php
<?php
$adapter->write($file, $data);
?>
```

#### Read data from a file: ####
```php
<?php
$adapter->read($file);
?>
```

#### Get a read-only stream resource for a file: ####
```php
<?php
$adapter->stream($file);
?>
```

#### Copy a file internally: ####
```php
<?php
$adapter->copy($src, $dest);
?>
```

#### Rename a file internally: ####
```php
<?php
$adapter->rename($src, $dest);
?>
```

#### Delete a file: ####
```php
<?php
$adapter->unlink($file);
?>
```

#### Check whether a file exists: ####
```php
<?php
$adapter->exists($file);
?>
```

#### Get a files size: ####
```php
<?php
$adapter->size($file);
?>
```

#### Get a files MIME content type: ####
```php
<?php
$adapter->type($file);
?>
```

#### Get the web-accessible uri for the given file: ####
```php
<?php
$adapter->uri($file);
?>
```

Adapters
--------

Cabinet offers two adapters:

  * `DotsUnited\Cabinet\Adapter\StreamAdapter` for [PHP streams](http://php.net/stream)
  * `DotsUnited\Cabinet\Adapter\AmazonS3Adapter` for [Amazon S3](https://s3.amazonaws.com)

Each adapter accepts its own set of configuration parameters which can be passed as an associative array to the constructor.

### DotsUnited\Cabinet\Adapter\StreamAdapter ###

Configuration parameters for `DotsUnited\Cabinet\Adapter\StreamAdapter`:

  * `base_path`:
     Path where to store the files.
  * `base_uri`:
     Uri where your files are publicly accessible.
  * `directory_umask`:
    The umask for directories created by the adapter (default is 0700).
  * `file_umask`:
    The umask for files created by the adapter (default is 0600).
  * `stream_context`:
    The [stream context](http://php.net/stream.contexts) to use with filesystem functions. This can be either a resource created with [`stream_context_create()`](http://php.net/stream-context-create) or an array with [context options](http://php.net/context).
  * `mime_type_detector`:
    An instance of `DotsUnited\Cabinet\MimeType\Detector\DetectorInterface` used to detect mime content types. This is optional, default is `DotsUnited\Cabinet\MimeType\Detector\Fileinfo`.
  * `filename_filter`:
    An instance of `DotsUnited\Cabinet\Filter\FilterInterface`. Filename filters are explained in the next section.

### DotsUnited\Cabinet\Adapter\AmazonS3Adapter ###

Configuration parameters for `DotsUnited\Cabinet\Adapter\AmazonS3Adapter`:

  * `s3_client`:
    A `Aws\S3\S3Client` instance (See the [AWS SDK docs](http://docs.aws.amazon.com/aws-sdk-php-2/guide/latest/service-s3.html#creating-a-client)).
  * `bucket`:
    Bucket where to store the files.
  * `storage_class`:
    The storage class setting for files. Allowed values: `STANDARD`, `REDUCED_REDUNDANCY`. The default value is `STANDARD`.
  * `acl`:
    The ACL settings for files. Allowed values: `private`, `public-read`, `public-read-write`, `authenticated-read`, `bucket-owner-read`, `bucket-owner-full-control`. The default value is `private`.
  * `uri_expiration_time`:
    The expiration time for web-accessible URIs if you store private files. This can bei either a timestamp or a string parsable by `strtotime()`.
  * `throw_exceptions`:
    A boolean indicating whether to throw exceptions (Exceptions from the `Aws\S3\S3Client` are catched and rethrown as `\RuntimeException`).
  * `mime_type_detector`:
    An instance of `DotsUnited\Cabinet\MimeType\Detector\DetectorInterface` used to detect mime content types. This is optional, default is `DotsUnited\Cabinet\MimeType\Detector\Fileinfo`.
  * `filename_filter`:
    An instance of `DotsUnited\Cabinet\Filter\FilterInterface`. Filename filters are explained in the next section.

Filename filters
----------------

You can manipulate the filename you pass to each method of an adapter with filters. Filters are classes which implement `DotsUnited\Cabinet\Filter\FilterInterface`.

You can add filters like this:

```php
<?php
$adapter->setFilenameFilter(new MyFilenameFilter());
?>
```

If you need multiple filters, you can use `DotsUnited\Cabinet\Filter\FilterChain` like this:

```php
<?php
$filterChain = new \DotsUnited\Cabinet\Filter\FilterChain();

$filterChain->addFilter(new MyFilenameFilter1());
$filterChain->addFilter(new MyFilenameFilter2());

$adapter->setFilenameFilter($filterChain);
?>
```

### Examples ###

Cabinet provides a filter to prepend a hashed subpath to the filename and is intended to be used with the `DotsUnited\Cabinet\Adapter\StreamAdapter` adapter.

This spreads the files over subdirectories to ensure performance by avoiding to many files in one directory. This is done by using a configurable number of characters of the filename's MD5 as a directory name. That pretty much guarantees an even spread.

Simply register the `DotsUnited\Cabinet\Filter\HashedSubpathFilter` filter with the adapter:

```php
<?php
$config = array(
    'level' => 4
);

$adapter->setFilenameFilter(new \DotsUnited\Cabinet\Filter\HashedSubpathFilter($config);
?>
```

License
-------

Cabinet is released under the [New BSD License](https://github.com/dotsunited/Cabinet/blob/master/LICENSE).


CHANGELOG
Changelog
=========

0.9.4 (2012-09-27)
------------------

  * Composer related fixes (moved AWS SDK to the suggest and used the packagist repository available since version 1.5.7)

0.9.3 (2012-02-28)
------------------

  * Make compatible with latest version 1.5.3 of aws/sdk (https://github.com/dotsunited/Cabinet/commit/d6a47ba7385bd61f40b2951e40d78544970d715d)

0.9.2 (2011-03-16)
------------------

  * Change all protected methods/properties to private (https://github.com/dotsunited/Cabinet/commit/ef81f3fdb3cded2c0bef3b948f0b138a2fa0cfa2)
  * Add default autoloader (https://github.com/dotsunited/Cabinet/commit/d31b4ad539ccc374a737885e5ae84c1ffc40e72e)
  * Fix deprecated PHPUnit methods (https://github.com/dotsunited/Cabinet/commit/cb819a33bd249b04e3149cd666a1ac3a2b2d4abc)

0.9.1 (2011-03-01)
------------------

  * Rename classes to comply with the Symfony2 conventions (https://github.com/dotsunited/Cabinet/commit/8f153b23e0948dc6ec059ad1ba50e0c892b458e7)
  * Remove short adapter class name (https://github.com/dotsunited/Cabinet/commit/a207849706927db05f18248482bdee76003437bf)

0.9.0 (2011-01-18)
------------------

  * Initial release


src/DotsUnited/Cabinet/Adapter/StreamAdapter.php
<?php

/*
 * This file is part of Cabinet.
 *
 * (c) 2012 Jan Sorgalla <jan.sorgalla@dotsunited.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace DotsUnited\Cabinet\Adapter;

use DotsUnited\Cabinet\Filter\FilterInterface;
use DotsUnited\Cabinet\MimeType\Detector\DetectorInterface;
use DotsUnited\Cabinet\MimeType\Detector\FileinfoDetector;

/**
 * DotsUnited\Cabinet\Adapter\StreamAdapter
 *
 * @author  Jan Sorgalla <jan.sorgalla@dotsunited.de>
 */
class StreamAdapter implements AdapterInterface
{
    /**
     * The base path.
     *
     * @var string
     */
    private $basePath;

    /**
     * The base uri.
     *
     * @var string
     */
    private $baseUri;

    /**
     * The directory umask.
     *
     * @var integer
     */
    private $directoryUmask = 0700;

    /**
     * The file umask.
     *
     * @var integer
     */
    private $fileUmask = 0600;

    /**
     * The stream context.
     *
     * @var array|resource
     */
    private $streamContext;

    /**
     * The mime type detector.
     *
     * @var \DotsUnited\Cabinet\MimeType\Detector\DetectorInterface
     */
    private $mimeTypeDetector;

    /**
     * The filename filter.
     *
     * @var \DotsUnited\Cabinet\Filter\FilterInterface
     */
    private $filenameFilter;

    /**
     * Constructor.
     *
     * @param array $config
     */
    public function __construct(array $config = array())
    {
        if (isset($config['base_path'])) {
            $this->setBasePath($config['base_path']);
        }

        if (isset($config['base_uri'])) {
            $this->setBaseUri($config['base_uri']);
        }

        if (isset($config['directory_umask'])) {
            $this->setDirectoryUmask($config['directory_umask']);
        }

        if (isset($config['file_umask'])) {
            $this->setFileUmask($config['file_umask']);
        }

        if (isset($config['stream_context'])) {
            $this->setStreamContext($config['stream_context']);
        }

        if (isset($config['mime_type_detector'])) {
            $this->setMimeTypeDetector($config['mime_type_detector']);
        }

        if (isset($config['filename_filter'])) {
            $this->setFilenameFilter($config['filename_filter']);
        }
    }

    /**
     * Set the base path.
     *
     * @param  string $basePath
     * @return Stream
     */
    public function setBasePath($basePath)
    {
        if (null !== $basePath) {
            if (substr($basePath, -3) != '://') {
                $basePath = rtrim($basePath, '/\\');
            }
        }

        $this->basePath = $basePath;

        return $this;
    }

    /**
     * Get the base path.
     *
     * @return string
     */
    public function getBasePath()
    {
        return $this->basePath;
    }

    /**
     * Set the base uri.
     *
     * @param  string $baseUri
     * @return Stream
     */
    public function setBaseUri($baseUri)
    {
        if (null !== $baseUri) {
            if (substr($baseUri, -3) != '://') {
                $baseUri = rtrim($baseUri, '/\\');
            }
        }

        $this->baseUri = $baseUri;

        return $this;
    }

    /**
     * Get the base uri.
     *
     * @return string
     */
    public function getBaseUri()
    {
        return $this->baseUri;
    }

    /**
     * Set the directory umask.
     *
     * @param  integer|string $directoryUmask
     * @return Stream
     */
    public function setDirectoryUmask($directoryUmask)
    {
        if (null !== $directoryUmask) {
            if (is_string($directoryUmask)) {
                $directoryUmask = octdec($directoryUmask);
            }
        }

        $this->directoryUmask = $directoryUmask;

        return $this;
    }

    /**
     * Get the directory umask.
     *
     * @return integer
     */
    public function getDirectoryUmask()
    {
        return $this->directoryUmask;
    }

    /**
     * Set the file umask.
     *
     * @param  integer|string $fileUmask
     * @return Stream
     */
    public function setFileUmask($fileUmask)
    {
        if (null !== $fileUmask) {
            if (is_string($fileUmask)) {
                $fileUmask = octdec($fileUmask);
            }
        }

        $this->fileUmask = $fileUmask;

        return $this;
    }

    /**
     * Get the file umask.
     *
     * @return integer
     */
    public function getFileUmask()
    {
        return $this->fileUmask;
    }

    /**
     * Set the stream context.
     *
     * @param  array|resource $streamContext
     * @return Stream
     */
    public function setStreamContext($streamContext)
    {
        $this->streamContext = $streamContext;

        return $this;
    }

    /**
     * Get the stream context.
     *
     * @return resource
     */
    public function getStreamContext()
    {
        if (null === $this->streamContext) {
            $this->streamContext = stream_context_create();
        } elseif (is_array($this->streamContext)) {
            $this->streamContext = stream_context_create($this->streamContext);
        }

        return $this->streamContext;
    }

    /**
     * Set the mime type detector.
     *
     * @param  \DotsUnited\Cabinet\MimeType\Detector\DetectorInterface $mimeTypeDetetcor
     * @return Stream
     */
    public function setMimeTypeDetector(DetectorInterface $mimeTypeDetetcor)
    {
        $this->mimeTypeDetector = $mimeTypeDetetcor;

        return $this;
    }

    /**
     * Get the mime type detector.
     *
     * @return \DotsUnited\Cabinet\MimeType\Detector\DetectorInterface
     */
    public function getMimeTypeDetector()
    {
        if (null === $this->mimeTypeDetector) {
            $this->setMimeTypeDetector(new FileinfoDetector());
        }

        return $this->mimeTypeDetector;
    }

    /**
     * Set the filename filter.
     *
     * @param \DotsUnited\Cabinet\Filter\FilterInterface
     * @return Stream
     */
    public function setFilenameFilter(FilterInterface $filter)
    {
        $this->filenameFilter = $filter;

        return $this;
    }

    /**
     * Get the filename filter.
     *
     * @return \DotsUnited\Cabinet\Filter\FilterInterface
     */
    public function getFilenameFilter()
    {
        return $this->filenameFilter;
    }

    /**
     * Import a external local file.
     *
     * @param  string  $external The external local file
     * @param  string  $file     The name to store the file under
     * @return boolean
     */
    public function import($external, $file)
    {
        $external = (string) $external;

        $path = $this->makePath($file);

        $ret = copy($external, $path, $this->getStreamContext());

        if ($ret) {
            @chmod($path, $this->getFileUmask());
        }

        return $ret;
    }

    /**
     * Write data to a file.
     *
     * @param  string                $file
     * @param  string|array|resource $data
     * @return boolean
     */
    public function write($file, $data)
    {
        $path = $this->makePath($file);

        $ret = (boolean) file_put_contents($path, $data, null, $this->getStreamContext());

        if ($ret) {
            @chmod($path, $this->getFileUmask());
        }

        return $ret;
    }

    /**
     * Read data from a file.
     *
     * @param  string         $file
     * @return string|boolean The contents or false on failure
     */
    public function read($file)
    {
        $path = $this->path($file);

        return file_get_contents($path, null, $this->getStreamContext());
    }

    /**
     * Return a read-only stream resource for a file.
     *
     * @param  string           $file
     * @return resource|boolean The resource or false on failure
     */
    public function stream($file)
    {
        $path = $this->path($file);

        return fopen($path, 'rb', false, $this->getStreamContext());
    }

    /**
     * Copy a file internally.
     *
     * @param  string  $src
     * @param  string  $dest
     * @return boolean Whether the file was copied
     */
    public function copy($src, $dest)
    {
        $srcPath  = $this->path($src);
        $destPath = $this->makePath($dest);

        $ret = copy($srcPath, $destPath, $this->getStreamContext());

        if ($ret) {
            @chmod($destPath, $this->getFileUmask());
        }

        return $ret;
    }

    /**
     * Rename a file internally.
     *
     * @param  string  $src
     * @param  string  $dest
     * @return boolean Whether the file was renamed
     */
    public function rename($src, $dest)
    {
        $srcPath  = $this->path($src);
        $destPath = $this->makePath($dest);

        $ret = rename($srcPath, $destPath, $this->getStreamContext());

        if ($ret) {
            @chmod($destPath, $this->getFileUmask());
            $this->deletePath($src);
        } else {
            $this->deletePath($dest);
        }

        return $ret;
    }

    /**
     * Delete a file.
     *
     * @param  string  $file
     * @return boolean Whether the file was deleted
     */
    public function unlink($file)
    {
        $path = $this->path($file);

        $ret = unlink($path, $this->getStreamContext());

        $this->deletePath($file);

        return $ret;
    }

    /**
     * Return whether a file exists.
     *
     * @param  string  $file
     * @return boolean Whether the file exists
     */
    public function exists($file)
    {
        $path = $this->path($file);

        return file_exists($path);
    }

    /**
     * Return the files size.
     *
     * @param  string  $file
     * @return integer The file size in bytes
     */
    public function size($file)
    {
        $path = $this->path($file);

        return filesize($path);
    }

    /**
     * Try to determine and return a files MIME content type.
     *
     * @param  string $file
     * @return string The MIME content type
     */
    public function type($file)
    {
        $path = $this->path($file);

        return $this->getMimeTypeDetector()->detectFromFile($path);
    }

    /**
     * Return the web-accessible uri for the given file.
     *
     * @param  string $file
     * @return string The file uri
     */
    public function uri($file)
    {
        if (null !== $this->filenameFilter) {
            $file = $this->filenameFilter->filter($file);
        }

        return $this->getBaseUri() . '/' .  str_replace(array('\\', '/'), '/', $file);
    }

    /**
     * Return the path for the given file.
     *
     * @param  string $file
     * @return string The file path
     */
    public function path($file)
    {
        $sep  = DIRECTORY_SEPARATOR;
        $path = $this->getBasePath();

        if (null !== $path) {
            if (strpos($path, '://') !== false) {
                $sep = '/';
            }

            $path .= $sep;
        }

        if (null !== $this->filenameFilter) {
            $file = $this->filenameFilter->filter($file);
        }

        return $path . str_replace(array('\\', '/'), $sep, $file);
    }

    /**
     * Return a file's parent.
     *
     * <code>null</code> is returned if the file has no parent.
     *
     * @param  string $file
     * @return string The parent directory of a file
     */
    private function parent($file)
    {
        $parent = dirname(rtrim($file, '/\\'));

        if (!empty($parent) && $parent != '.') {
            return $parent;
        }

        return null;
    }

    /**
     * Create a file path for the given file.
     *
     * @param  string         $file
     * @return string|boolean The path on success, false otherwise
     */
    private function makePath($file)
    {
        $path     = $this->path($file);
        $dir      = $this->parent($path);
        $basePath = $this->getBasePath();

        if (null !== $dir && $dir !== $basePath) {
            $dir  = substr($dir, strlen($basePath));

            $parts = preg_split('/[\\\\\/]+/', trim($dir, '/\\'));

            $dirUmask      = $this->getDirectoryUmask();
            $dirPath       = $basePath;
            $streamContext = $this->getStreamContext();

            foreach ($parts as $part) {
                $dirPath .= DIRECTORY_SEPARATOR . $part;
                if (!is_dir($dirPath)) {
                    mkdir($dirPath, $dirUmask, false, $streamContext);
                    @chmod($dirPath, $dirUmask); // Required in some configurations
                }
            }
        }

        return $path;
    }

    /**
     * Delete the internal path for a file.
     *
     * @param  string $file
     * @return void
     */
    private function deletePath($file)
    {
        $path          = $this->path($file);
        $parent        = $this->parent($path);
        $basePath      = $this->getBasePath();
        $streamContext = $this->getStreamContext();

        while (true) {
            if (null === $parent || $parent === $basePath || !file_exists($parent)) {
                break;
            }

            if (count(scandir($parent, null, $streamContext)) > 2) { // Folder contains files
                break;
            }

            if (!rmdir($parent, $streamContext)) { // Something went wrong
                break;
            }

            $parent = $this->parent($parent);
        }
    }
}
